src/models/dirac_mixture_single_publisher_model.py

```diff
@@ -0,0 +1,246 @@
+        """Sampling grid according to a pmf.
+
+        For each frequency level, sample n_f points equally spaced in
+        [f, f + 1], where n_f (before rounding) is proportional to pmf[f].
+
+        Args:
+            ncomponents: Number of components to sample.
+            pmf: A vector of probabilities that sum up to be 1.
+
+        Returns:
+            All components in the model.
+        """
+        n_left = ncomponents
+        components = np.array([])
+        for f in range(len(pmf)):
+            n = int(ncomponents * pmf[f])
+            components = np.concatenate(
+                (components, np.linspace(f, f + 1, n, endpoint=False))
+            )
+            n_left -= n
+        if n_left > 0:
+            # Due to rounding, there can remain a certain number of components.
+            # Sample these remaining components in the next interval,
+            # that is, [F + 1, F + 2], where F is the maximum frequency.
+            components = np.concatenate(
+                (
+                    components,
+                    np.linspace(len(pmf), len(pmf) + 1, n_left, endpoint=False),
+                )
+            )
+        return components
+
+    @classmethod
+    def truncated_poisson_pmf_vec(
+        cls, poisson_mean: float, max_freq: int
+    ) -> np.ndarray:
+        """pmf vector of a Poisson distribution truncated at a max frequency.
+
+        Args:
+            poisson_mean: mean of Poisson distribution.
+            max_freq: Any observation beyond this maximum frequency will be
+                rounded to this maximum frequency.
+
+        Returns:
+            A vector v where v[f] = Poisson_pmf at f for f < max_freq, and
+            v[max_freq] = (probability >= max_freq)
+            = 1 - Poisson_cdf(max_freq - 1).
+        """
+        v = stats.poisson.pmf(range(max_freq + 1), poisson_mean)
+        v[-1] = 1 - stats.poisson.cdf(max_freq - 1, poisson_mean)
+        return v
+
+    @classmethod
+    def cross_entropy(cls, observed_arr: np.ndarray, fitted_arr: np.ndarray) -> float:
+        """The cross entropy distance between observed and fitted arrays.
+
+        Minimizing cross entropy is equivalent to maximizing likelihood.
+
+        Args:
+            observed_arr: A 1d array of the observed pmf. Or 2d array of the
+                observed pmfs on different directions.
+            fitted_arr: Any other array of pmf(s) that has the same dimension
+                as observed_arr.
+        Returns:
+             The cross entropy of the fitted pmf(s) relative to the observed
+             pmf(s), that is,
+            - sum_i observed_arr[i] * log(fitted_arr[i]) for 1d arrays,
+            - sum_{i, j} observed_arr[i, j] * log(fitted_arr[i, j]) for 2d arrays.
+        """
+        # Small shift to avoid log-zero numerical error.
+        fitted_arr = fitted_arr + 1e-200
+        return -cp.sum(cp.multiply(observed_arr, cp.log(fitted_arr)))
```

Comment:
Is it required to use cvxpy operations in the objective?

Reply:
Good question, I feel yes.  Because the cvxpy problem (in the next method) will call this method as objective function.